src/ast/value.rs

```diff
@@ -148,47 +164,67 @@ pub enum DateTimeField {
     TimezoneMinute,
     TimezoneRegion,
     NoDateTime,
+    /// Arbitrary abbreviation or custom date-time part.
+    ///
+    /// ```sql
+    /// EXTRACT(q FROM CURRENT_TIMESTAMP)
+    /// ```
+    /// [Snowflake](https://docs.snowflake.com/en/sql-reference/functions-date-time#supported-date-and-time-parts)
+    Custom(Ident),
 }
 
 impl fmt::Display for DateTimeField {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        f.write_str(match self {
-            DateTimeField::Year => "YEAR",
-            DateTimeField::Month => "MONTH",
-            DateTimeField::Week => "WEEK",
-            DateTimeField::Day => "DAY",
-            DateTimeField::DayOfWeek => "DAYOFWEEK",
-            DateTimeField::DayOfYear => "DAYOFYEAR",
-            DateTimeField::Date => "DATE",
-            DateTimeField::Hour => "HOUR",
-            DateTimeField::Minute => "MINUTE",
-            DateTimeField::Second => "SECOND",
-            DateTimeField::Century => "CENTURY",
-            DateTimeField::Decade => "DECADE",
-            DateTimeField::Dow => "DOW",
-            DateTimeField::Doy => "DOY",
-            DateTimeField::Epoch => "EPOCH",
-            DateTimeField::Isodow => "ISODOW",
-            DateTimeField::Isoyear => "ISOYEAR",
-            DateTimeField::IsoWeek => "ISOWEEK",
-            DateTimeField::Julian => "JULIAN",
-            DateTimeField::Microsecond => "MICROSECOND",
-            DateTimeField::Microseconds => "MICROSECONDS",
-            DateTimeField::Millenium => "MILLENIUM",
-            DateTimeField::Millennium => "MILLENNIUM",
-            DateTimeField::Millisecond => "MILLISECOND",
-            DateTimeField::Milliseconds => "MILLISECONDS",
-            DateTimeField::Nanosecond => "NANOSECOND",
-            DateTimeField::Nanoseconds => "NANOSECONDS",
-            DateTimeField::Quarter => "QUARTER",
-            DateTimeField::Time => "TIME",
-            DateTimeField::Timezone => "TIMEZONE",
-            DateTimeField::TimezoneAbbr => "TIMEZONE_ABBR",
-            DateTimeField::TimezoneHour => "TIMEZONE_HOUR",
-            DateTimeField::TimezoneMinute => "TIMEZONE_MINUTE",
-            DateTimeField::TimezoneRegion => "TIMEZONE_REGION",
-            DateTimeField::NoDateTime => "NODATETIME",
-        })
+        f.write_str(
+            match self {
+                DateTimeField::Year => "YEAR".to_string(),
```

Comment:
This change to call `to_string()` will now cause a new string to be allocated for each type.

I think it would be better / more effiicent to use `write!` directly -- like

```suggestion
              match self {
                DateTimeField::Year => f.write_str("YEAR"),
                DateTimeField::Month => f.write_str("MONTH"),
                ...
```

Reply:
Ah yes that makes a lot more sense, thanks for the follow up fix too!

Reply:
(for anyone following along it was https://github.com/sqlparser-rs/sqlparser-rs/pull/1209)